src/Text/Pretty/Simple/Internal/OutputPrinter.hs

```diff
@@ -121,81 +130,116 @@ hCheckTTY h options = liftIO $ conv <$> tty
-    , useColorString
-    , pure (fromString string)
-    , useColorReset
-    , useColorQuote
-    , pure "'"
-    , useColorReset
-    ]
+    list open close (CommaSeparated es) = annotate AnnOpen open <>
+        hcat (intersperse (line <> annotate AnnComma ",")
+          (map (exprs True) es)) <> l <> annotate AnnClose close
+      where l = if null es then line' else line
+    exprs nested = \case -- 'nested' is True unless this is the very top level
+      [] -> mempty
+      x : xs -> expr nested x <> hcat (map (expr True) xs)
+    expr nested x = applyWhen (isSimple x) group $
+      applyWhen nested ((line <>) . flatAlt (indent indentAmount (go x))) $ go x
+    isSimple = \case -- should this expression be contracted on to one line?
+      (getList -> Just [[e]]) -> isSimple e
+      (getList -> Just (_:_)) -> False
+      StringLit s -> '\n' `notElem` s
+      _ -> True
+    getList = \case
+      Brackets (CommaSeparated xs) -> Just xs
+      Braces (CommaSeparated xs) -> Just xs
+      Parens (CommaSeparated xs) -> Just xs
+      _ -> Nothing
+    go = \case
+      Brackets xss -> list "[" "]" xss
+      Braces xss -> list "{" "}" xss
+      Parens xss -> list "(" ")" xss
+      StringLit s -> join enclose (annotate AnnQuote "\"") $ annotate AnnString $ pretty $
+        applyWhen (outputOptionsEscapeNonPrintable opts) escapeNonPrintable s
+      CharLit s -> join enclose (annotate AnnQuote "'") $ annotate AnnString $ pretty s
+      Other s -> pretty s
+      NumberLit n -> annotate AnnNum $ pretty n
+    indentAmount = outputOptionsIndentAmount opts
+
+preprocess :: [Expr] -> [Expr]
+preprocess = map one . removeEmptyOther
+  where
+    one = \case
+      Brackets xss -> Brackets $ cs xss
+      Braces xss -> Braces $ cs xss
+      Parens xss -> Parens $ cs xss
+      StringLit s -> StringLit $ read $ "\"" <> s <> "\""
+      CharLit s -> CharLit s
+      Other s -> Other $ shrinkWhitespace $ strip s
+      NumberLit n -> NumberLit n
+    cs (CommaSeparated ess) = CommaSeparated $ map preprocess ess
+
+removeEmptyOther :: [Expr] -> [Expr]
+removeEmptyOther = filter $ \case
+  Other s -> any (not . isSpace) s
+  _ -> True
+
+applyWhen :: Bool -> (a -> a) -> a -> a
+applyWhen = flip $ bool id
+
+annotateAnsi :: OutputOptions -> SimpleDocStream Ann
+  -> SimpleDocStream AnsiStyle
+annotateAnsi opts = case outputOptionsColorOptions opts of
+  Nothing -> unAnnotateS
+  Just co -> flip evalState (newD (colorError co) rainbowParensColor) .
+    traverse f
+    where
+      f = \case
+        AnnOpen -> state advanceD >> gets getD
+        AnnClose -> state $ retreatD
+        AnnComma -> gets getD
+        AnnQuote -> pure $ colorQuote co
+        AnnString -> pure $ colorString co
+        AnnNum -> pure $ colorNum co
+      rainbowParensColor = fromMaybe (pure mempty) $ NE.nonEmpty $
+        colorRainbowParens co
+
+data Ann
+  = AnnOpen
+  | AnnClose
+  | AnnComma
+  | AnnQuote
+  | AnnString
+  | AnnNum
+
+data D a = D
```

Comment:
This datatype gives us a nice, elegant, and provably total implementation of `annotateAnsi`, but it incurs a dependency on `Stream` we don't really want. We can definitely implement something simpler with ordinary lists instead (and also think of a proper name for it...).

Reply:
It looks fascinating, but I have no idea what's going on! ;)

Some type annotations and comments in `annotateAnsi` would really help.

Reply:
This should be a lot clearer now. We use an infinite `Tape` of `AnsiStyle`s, on which we move right at an open bracket, and left at a close.